File: src/routes/valuation.ts
/**
 * Valuation Routes
 *
 * Endpoints for AVM (Automated Valuation Model) from DataTree
 */

import type { FastifyInstance, FastifyRequest, FastifyReply } from 'fastify';
import { dataTreeAVM } from '../adapters/datatree/DataTreeAdapter.js';

interface AddressBody {
  street: string;
  unit?: string;
  city: string;
  state: string;
  zipCode: string;
}

interface AVMOrderBody {
  applicationId?: string;
  propertyId?: string;
  address: AddressBody;
}

export async function valuationRoutes(fastify: FastifyInstance) {
  /**
   * POST /valuation/avm
   * Order an AVM for a property
   */
  fastify.post('/avm', async (
    request: FastifyRequest<{ Body: AVMOrderBody }>,
    reply: FastifyReply
  ) => {
    const { applicationId, propertyId, address } = request.body;

    if (!address || !address.street || !address.city || !address.state || !address.zipCode) {
      return reply.status(400).send({
        success: false,
        error: {
          code: 'INVALID_ADDRESS',
          message: 'Missing required address fields: street, city, state, zipCode',
        },
      });
    }

    const result = await dataTreeAVM.orderAVM(address);

    if (!result.success) {
      return reply.status(422).send({
        success: false,
        orderId: result.orderId,
        error: result.error,
      });
    }

    // Attach applicationId and propertyId if provided
    if (result.report) {
      result.report.applicationId = applicationId || '';
      result.report.propertyId = propertyId || '';
    }

    return {
      success: true,
      orderId: result.orderId,
      report: result.report,
    };
  });

  /**
   * GET /valuation/avm/:orderId
   * Get AVM report by order ID (for async workflows)
   */
  fastify.get('/avm/:orderId', async (
    request: FastifyRequest<{ Params: { orderId: string } }>,
    reply: FastifyReply
  ) => {
    const { orderId } = request.params;

    const report = await dataTreeAVM.getReport(orderId);

    if (!report) {
      return reply.status(404).send({
        error: 'AVM report not found',
      });
    }

    return report;
  });

  /**
   * POST /valuation/quick-value
   * Quick property value lookup (simplified AVM response)
   */
  fastify.post('/quick-value', async (
    request: FastifyRequest<{ Body: { address: AddressBody } }>,
    reply: FastifyReply
  ) => {
    const { address } = request.body;

    if (!address || !address.street || !address.city || !address.state || !address.zipCode) {
      return reply.status(400).send({
        error: 'Missing required address fields',
      });
    }

    const result = await dataTreeAVM.orderAVM(address);

    if (!result.success || !result.report) {
      return {
        hasValue: false,
        error: result.error?.message || 'No value available',
      };
    }

    const report = result.report;

    return {
      hasValue: true,
      value: {
        estimated: report.estimatedValue ? report.estimatedValue / 100 : null,
        low: report.valueLow ? report.valueLow / 100 : null,
        high: report.valueHigh ? report.valueHigh / 100 : null,
      },
      confidence: {
        score: report.confidenceScore,
        level: report.confidenceLevel,
      },
      property: report.propertyCharacteristics ? {
        type: report.propertyCharacteristics.propertyType,
        yearBuilt: report.propertyCharacteristics.yearBuilt,
        sqft: report.propertyCharacteristics.squareFeet,
        beds: report.propertyCharacteristics.bedrooms,
        baths: report.propertyCharacteristics.bathrooms,
      } : null,
      lastSale: report.lastSaleDate ? {
        date: report.lastSaleDate,
        price: report.lastSalePrice ? report.lastSalePrice / 100 : null,
      } : null,
      comparablesCount: report.comparables?.length || 0,
      vendor: report.vendorName,
    };
  });
}

File: api/app/routers/valuation.py
"""Valuation router - AVM and appraisal endpoints."""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from app.adapters.datatree import Address
from app.services.valuation import valuation_service

router = APIRouter()


class ValuationRequest(BaseModel):
    """Valuation request."""
    application_id: str
    street: str
    city: str
    state: str
    zip_code: str
    unit: str | None = None


class AVMResponse(BaseModel):
    """AVM response."""
    order_id: str
    status: str
    estimated_value: int | None  # cents
    confidence: str | None
    vendor: str | None
    value_low: int | None
    value_high: int | None
    error_message: str | None = None


class ReconciliationRequest(BaseModel):
    """Value reconciliation request."""
    avm_value: int  # cents
    appraisal_value: int  # cents
    tolerance: float = 0.10


class ReconciliationResponse(BaseModel):
    """Value reconciliation response."""
    reconciled: bool
    avm_value: int
    appraisal_value: int
    variance_pct: float
    variance_direction: str
    within_tolerance: bool
    recommended_value: int


@router.post("/avm", response_model=AVMResponse)
async def order_avm(request: ValuationRequest) -> AVMResponse:
    """
    Order an AVM (Automated Valuation Model) for a property.
    Uses cascade logic to try multiple vendors.
    """
    address = Address(
        street=request.street,
        city=request.city,
        state=request.state,
        zip_code=request.zip_code,
        unit=request.unit,
    )

    order = await valuation_service.order_avm(request.application_id, address)

    return AVMResponse(
        order_id=order.id,
        status=order.status.value,
        estimated_value=order.estimated_value,
        confidence=None,  # Would come from the report
        vendor=order.vendor_name,
        value_low=None,
        value_high=None,
        error_message=order.error_message,
    )


@router.get("/avm/{order_id}")
async def get_avm(order_id: str) -> dict:
    """Get AVM order status and results."""
    order = await valuation_service.get_valuation(order_id)

    if not order:
        raise HTTPException(status_code=404, detail="AVM order not found")

    return {
        "order_id": order.id,
        "status": order.status.value,
        "estimated_value": order.estimated_value,
        "vendor": order.vendor_name,
        "order_date": order.order_date.isoformat(),
        "completed_date": order.completed_date.isoformat() if order.completed_date else None,
    }


@router.post("/reconcile", response_model=ReconciliationResponse)
async def reconcile_values(request: ReconciliationRequest) -> ReconciliationResponse:
    """
    Reconcile AVM and appraisal values.
    Returns variance analysis and recommended value.
    """
    result = valuation_service.reconcile_values(
        avm_value=request.avm_value,
        appraisal_value=request.appraisal_value,
        tolerance=request.tolerance,
    )

    if "error" in result:
        raise HTTPException(status_code=400, detail=result["error"])

    return ReconciliationResponse(
        reconciled=result["reconciled"],
        avm_value=result["avm_value"],
        appraisal_value=result["appraisal_value"],
        variance_pct=result["variance_pct"],
        variance_direction=result["variance_direction"],
        within_tolerance=result["within_tolerance"],
        recommended_value=result["recommended_value"],
    )
